Add Root and Halo2 Prover/Verifier

darth-cy opened the pull request:
- Write Ceno's independent root prover.
- Use the new internal proving logic.

For halo2 proof, run test test_static_verifier
- To test root AIR permutation using trace heights, use test test_root_proof_permutation
- Before running above test, run test_aggregation to generate a root proof.
- Download KZG params (default k=23) from these links: https://github.com/scroll-tech/openvm/blob/feat/v1.4.1-scroll-ext/docs/vocs/docs/pages/specs/security/security-model.mdx

darth-cy committed:
add root prover [skip ci]

darth-cy committed:
use self-defined root verifier input [skip ci]

darth-cy committed:
use custom internal [skip ci]

darth-cy committed:
Merge branch 'master' into feat/root_prover_verifier

darth-cy committed:
ignore test

darth-cy committed:
clippy

darth-cy committed:
Merge branch 'master' into feat/root_prover_verifier

darth-cy committed:
Merge branch 'master' into feat/root_prover_verifier

darth-cy committed:
add init_pc to ceno root pv

darth-cy committed:
correct root pv len

darth-cy committed:
New Recursion PV + EC Sum Aggregation/Check (Internal & Root) (#1196)

- Restructure the public values in recursion
- Aggregate and check septic point summation in the aggregation tree

darth-cy committed:
Merge branch 'master' into feat/root_prover_verifier

darth-cy committed:
ignore json [skip ci]

darth-cy committed:
verify root proof [skip ci]

darth-cy committed:
verify root proof

hero78119 reviewed (changes requested):
Try new api in this PR on https://github.com/scroll-tech/ceno-reth-benchmark e2e, need to adjust further of new api design

darth-cy committed:
verify root proof and restore sdk

darth-cy commented on ceno_recursion/src/aggregation/mod.rs:
```diff
@@ -356,13 +405,35 @@ impl CenoAggregationProver {
+        root_proof
+    }
+
+    pub fn verify_root_proof(&self, root_proof: &Proof<RootSC>) -> Result<(), VerificationError> {
```
A separate verification function is added

darth-cy commented on ceno_cli/src/sdk.rs:
```diff
@@ -194,25 +191,12 @@ where
-                &agg_pk.leaf_vm_pk,
-                agg_pk.leaf_committed_exe.exe.clone(),
+        let mut agg_prover = if let Some(_agg_pk) = self.agg_pk.as_ref() {
+            CenoAggregationProver::from_base_vk(
```
Restored the original initiation

hero78119 committed:
misc refactor

hero78119 reviewed (commented):
Reported with e2e test with block `23587691`
- e2e prove/verify pass
- final root proof size: 671K
- currently only cpu mode, as I didn't find respective gpu config for root prover. On cpu root proof take `89.78s`. We need a way to switch to gpu config

darth-cy committed:
Merge branch 'master' into feat/root_prover_verifier

darth-cy committed:
Merge branch 'master' into feat/root_prover_verifier

darth-cy committed:
Merge branch 'master' into feat/root_prover_verifier

darth-cy committed:
fmt

darth-cy committed:
Merge branch 'master' into feat/root_prover_verifier

darth-cy committed:
Merge branch 'master' into feat/root_prover_verifier

darth-cy committed:
Add Halo2 Proof to Recursion (#1236)

- Run test `test_static_verifier`
- To test root AIR permutation using trace heights, use test
`test_root_proof_permutation`
- Before running above test, run `test_aggregation` to generate a root
proof.
- Download KZG params (default k=23) from these links:
https://github.com/scroll-tech/openvm/blob/feat/v1.4.1-scroll-ext/docs/vocs/docs/pages/specs/security/security-model.mdx

darth-cy committed:
Merge branch 'master' into feat/root_prover_verifier

darth-cy committed:
update cargo

darth-cy committed:
restore tests

darth-cy committed:
fmt

darth-cy committed:
keep params and exports dir

darth-cy committed:
download kzg params in integration

darth-cy committed:
add conditional in ci

darth-cy committed:
remove intermediate proof exports

darth-cy committed:
remove std file [skip ci]

darth-cy committed:
path buff [skip ci]

darth-cy committed:
correct path [skip ci]

darth-cy committed:
correct path [skip ci]

darth-cy committed:
remove verifier gen [skip ci]

darth-cy committed:
export intermediate proofs

darth-cy committed:
Merge branch 'master' into feat/root_prover_verifier
